File: reverse.js
// https://leetcode.com/problems/reverse-integer/
// 
// Given a signed 32-bit integer x, return x with its digits reversed. If reversing x causes the value to go outside the signed 32-bit integer range [-231, 231 - 1], then return 0.
// 
// Assume the environment does not allow you to store 64-bit integers (signed or unsigned).
// 
//  
// Example 1:
// Input: x = 123
// Output: 321
//
// Example 2:
// Input: x = -123
// Output: -321
//
// Example 3:
// Input: x = 120
// Output: 21



/**
 * @param {number} x
 * @return {number}
 */
var reverse = function (x) {

    var reversed = parseInt(String(Math.abs(x)).split('').reverse().join(''));

    if (x < 0) {
        reversed = - reversed;
    }

    if (reversed < Math.pow(2,31)-1 && reversed > -Math.pow(2,31)) {
        return reversed
    }

    return 0;
};

// Runtime: 64 ms, faster than 99.25% of JavaScript online submissions for Reverse Integer.
// Memory Usage: 43.9 MB, less than 34.36% of JavaScript online submissions for Reverse Integer.

module.exports = {
  reverse,
};

File: leetcode/Algorithms/7-Reverse-Integer/Reverse.cs
/*
    Source: https://leetcode.com/problems/reverse-integer
    Runtime: 40 ms, faster than 32.22% of C# online submissions for Reverse Integer.
    Memory Usage: 25.5 MB, less than 64.97% of C# online submissions for Reverse Integer.
    Author: berkslv
    Date: 14/06/2022
*/
namespace Algorithms
{
    public class Solution007
    {
        public static int Reverse(int x)
        {
            if (x <= int.MinValue || x >= int.MaxValue)
            {
                return 0;
            }

            var integer = Math.Abs(x);
            var result = 0;
            
            while (integer != 0)
            {
                if (result > (int.MaxValue - integer % 10) / 10)
                {
                    return 0;
                }
            
                result = result * 10 + integer % 10;
                integer /= 10;
            
            }
            
            return x > 0 ? result : -result;
        }
    }
}
